### PR TITLE
feat: Support custom delay time for scheduled tasks

### PR DESCRIPTION
**Pull request by pangdayuan1**

### CODE REVIEW

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanMonitorHandler.java [hunk 1]
```diff
@@ -17,4 +17,10 @@ public interface PlanMonitorHandler {
    * Execute when deRegister.
    */
   void end(ReplayPlan plan);
+
+  /**
+   * Get the delay time.
+   */
+  long getDelayTime();
```
Comment (mr3) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanMonitorHandler.java:24]: `default long period() {
    return xxx;
}`

**Review by wildeslam — DISMISSED** [review]

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java [hunk 2]
```diff
@@ -34,52 +32,56 @@ public void monitorOne(ReplayPlan task) {
...
-      } catch (Throwable t) {
-        LOGGER.error("Error handling plan:{}", task.getId(), t);
+      } catch (Exception e) {
+        LOGGER.error("Error handling plan:{}", task.getId(), e);
       }
     }
   }
 
   @Override
   public void register(ReplayPlan plan) {
```
Comment (mr3) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java:42]: What if I have duplicate registrations?

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java [hunk 3]
```diff
@@ -34,52 +32,56 @@ public void monitorOne(ReplayPlan task) {
...
+    }
+    plan.setMonitorScheduler(monitorScheduler);
   }
 
   @Override
   public void deregister(ReplayPlan plan) {
-    plan.getMonitorFuture().cancel(false);
     LOGGER.info("deregister monitor task, planId: {}", plan.getId());
+    ScheduledExecutorService monitorExecutor = plan.getMonitorScheduler();
+    monitorExecutor.shutdown();
```
Comment (mr3) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java:54]: How do other replay plan tasks run after shutdown?

**Review by mr3 — APPROVED** [review]

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java (hunk 2 above)
Comment (pangdayuan1) [reply arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java:42]: The monitorScheduler is injected through the @Resource annotation and is a singleton. Actually more tasks are scheduled on the same ScheduledExecutorService instance (i.e. monitorScheduler), which means these tasks will share the same thread pool.

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java (hunk 3 above)
Comment (pangdayuan1) [reply arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java:54]: This is shutdown in the plan dimension and will not affect other replay plans.

Comment (mr3) [reply arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java:54]: Each replay plan monitor task share the same thread pool: `monitorScheduler`, if this executor shutdown, the other monitor tasks will stop executing as well

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java (hunk 2 above)
Comment (mr3) [reply arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java:42]: If a replayPlan is re-registered multiple times, duplicate MonitorTasks are created

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java [hunk 4]
```diff
@@ -19,70 +21,60 @@
...
+      return;
+    }
 
-    for (PlanMonitorHandler handler : planMonitorHandlerList) {
-      try {
+    try {
+      LOGGER.info("deregister monitor task, planId: {}", plan.getId());
+      monitorFutures.forEach(future -> future.cancel(false));
+      for (PlanMonitorHandler handler : planMonitorHandlerList) {
         handler.end(plan);
```
Comment (mr3) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java:73]: Here you need to catch exceptions to ensure that other handlers execute properly.

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java [hunk 5]
```diff
@@ -19,70 +21,60 @@
...
+                                   @Override
+                                   protected void doWithTracedRunning() {
+                                     try {
+                                       handler.handle(plan);
+                                     } catch (Exception e) {
+                                       LOGGER.error("failed to handle monitor task. plan:{}", plan.getId(), e);
+                                     }
+                                   }
+                                 }, 0, handler.period(),
+                TimeUnit.SECONDS);
```
Comment (mr3) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/impl/PlanExecutionMonitorImpl.java:53]: format as follows
```java
      List<ScheduledFuture<?>> monitorFutures = Lists.newArrayListWithCapacity(planMonitorHandlerList.size());

      for (PlanMonitorHandler handler : planMonitorHandlerList) {
        ScheduledFuture<?> monitorFuture = monitorScheduler.scheduleAtFixedRate(new AbstractTracedRunnable() {
          @Override
          protected void doWithTracedRunning() {
            try {
              handler.handle(plan);
            } catch (Exception e) {
              LOGGER.error("failed to handle monitor task. plan:{}", plan.getId(), e);
            }
          }
        }, 0, handler.period(), TimeUnit.SECONDS);
        monitorFutures.add(monitorFuture);
      }
```

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanMonitorHandler.java [hunk 6]
```diff
@@ -17,4 +18,12 @@ public interface PlanMonitorHandler {
    * Execute when deRegister.
    */
   void end(ReplayPlan plan);
+
+  /**
+   * Get period.
+   */
+  default long period() {
+    return DEFAULT_DELAY_SECOND;
+  }
```
Comment (mr3) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanMonitorHandler.java:27]:
```java
default long periodSeconds() {
    return 5L;
}
```

**Review by mr3 — DISMISSED** [review]

**Review by QizhengMo — DISMISSED** [review]

**Review by mr3 — APPROVED** [review]

**Review by QizhengMo — APPROVED** [review]

**Pull request merged**